scripts/build_test.py

```diff
@@ -101,3 +396,102 @@ def test_get_hashes_json_file_contents(self):
+                args=('path/to/input.js', 'path/to/output.js'))
+            build_tasks.append(task)
+            count -= 1
+
+        self.assertEqual(threading.active_count(), 1)
+        build._execute_tasks(build_tasks)
+        with self.assertRaises(OSError) as threadAlreadyStarted:
+            build._execute_tasks(build_tasks)
+        self.assertTrue(
+            'threads can only be started once' in
+            threadAlreadyStarted.exception)
+        # Assert that all threads are joined.
+        self.assertEqual(threading.active_count(), 1)
+
+    def test_generate_build_tasks_to_build_all_files_in_directory(self):
+        """Test generate_build_tasks_to_build_all_files_in_directory to queue up
+        the same number of build tasks as the number of files in the source
+        directory.
+        """
+        asset_hashes = build.get_file_hashes(build.ASSETS_DEV_DIR)
+        tasks = collections.deque()
+
+        self.assertEqual(len(tasks), 0)
+        # Build all files.
+        tasks = build.generate_build_tasks_to_build_all_files_in_directory(
+            build.ASSETS_DEV_DIR, build.ASSETS_OUT_DIR, asset_hashes)
+        total_file_count = build.get_file_count(build.ASSETS_DEV_DIR)
+        self.assertEqual(len(tasks), total_file_count)
+
+    def test_generate_build_tasks_to_build_files_from_filepaths(self):
+        """Test generate_build_tasks_to_build_files_from_filepaths queue up a
+        corresponding number of build tasks to the number of file changes.
+        """
+        new_filename = 'manifest.json'
+        recently_changed_filenames = [
+            os.path.join(build.ASSETS_DEV_DIR, 'i18n', new_filename)]
+        asset_hashes = build.get_file_hashes(build.ASSETS_DEV_DIR)
+        build_tasks = collections.deque()
+
+        self.assertEqual(len(build_tasks), 0)
+        build_tasks += build.generate_build_tasks_to_build_files_from_filepaths(
+            build.ASSETS_DEV_DIR, build.ASSETS_OUT_DIR,
+            recently_changed_filenames, asset_hashes)
+        self.assertEqual(len(build_tasks), len(recently_changed_filenames))
+
+        build_tasks.clear()
+        svg_filepaths = build.get_filepaths_by_extensions(
+            build.ASSETS_DEV_DIR, ('.svg',))
+        # Make sure there is at least 1 SVG file.
+        self.assertGreater(len(svg_filepaths), 0)
+
+        self.assertEqual(len(build_tasks), 0)
+        build_tasks += build.generate_build_tasks_to_build_files_from_filepaths(
+            build.ASSETS_DEV_DIR, build.ASSETS_OUT_DIR,
+            svg_filepaths, asset_hashes)
+        self.assertEqual(len(build_tasks), len(svg_filepaths))
+
+
+    def test_get_recently_changed_filenames(self):
+        """Test get_recently_changed_filenames to detect file recently added."""
+        assets_hashes = build.get_file_hashes(build.ASSETS_DEV_DIR)
+        # Create an empty sandbox folder, to simulate an empty /build folder.
+        SANDBOX_DIR = os.path.join(build.PARENT_DIR, 'sandbox')
```

Comment:
Suggest just using a (nuked) test dir in core/tests/build

Reply:
I don't think I need to access to a real folder here. Anyhow, this function now operates from within core/tests/build.